File: lib/engine/addOns.ts
/** Add-on acquisition modeling — injects bolt-on revenue/EBITDA into projections
 *  and bolt-on acquisition debt into the debt schedule. */

import type { ModelState, DebtTranche } from '../dealEngineTypes';

export interface AddOnImpact {
  /** Additional revenue injected per year (cumulative from all acquisitions closed by that year). */
  revenue_by_year: number[];
  /** Additional EBITDA injected per year. */
  ebitda_by_year: number[];
  /** Total acquisition cost funded by equity (affects IRR timing). */
  equity_deployed_by_year: number[];
  /** Total acquisition cost funded by debt. */
  debt_added_by_year: number[];
  /** One-time integration costs by year. */
  integration_cost_by_year: number[];
  /** Synergy revenue by year (cumulative). */
  synergy_revenue_by_year: number[];
  /** Cost synergies by year (cumulative). */
  cost_synergies_by_year: number[];
}

export function computeAddOnImpact(state: ModelState): AddOnImpact {
  const hp = state.exit.holding_period;
  const addOns = state.add_on_acquisitions || [];

  const revenueByYear = Array(hp).fill(0);
  const ebitdaByYear = Array(hp).fill(0);
  const equityByYear = Array(hp).fill(0);
  const debtByYear = Array(hp).fill(0);
  const integrationByYear = Array(hp).fill(0);
  const synergyRevByYear = Array(hp).fill(0);
  const costSynByYear = Array(hp).fill(0);

  for (const addon of addOns) {
    const acqYear = addon.year; // 1-indexed
    if (acqYear < 1 || acqYear > hp) continue;

    const purchasePrice = addon.revenue * addon.ebitda_margin * addon.purchase_multiple;
    const debtPortion = addon.funding === 'debt' ? purchasePrice
      : addon.funding === 'equity' ? 0
      : purchasePrice * addon.debt_pct;
    const equityPortion = purchasePrice - debtPortion;

    // Costs in acquisition year
    const yrIdx = acqYear - 1;
    equityByYear[yrIdx] += equityPortion;
    debtByYear[yrIdx] += debtPortion;
    integrationByYear[yrIdx] += addon.integration_cost;

    // Revenue and EBITDA from acquisition year onwards (assume full year from year of acquisition)
    for (let i = yrIdx; i < hp; i++) {
      const yearsOwned = i - yrIdx;
      // Simple: add-on revenue grows at the same rate as the base business
      const addOnGrowth = yearsOwned > 0 && i < state.revenue.growth_rates.length
        ? state.revenue.growth_rates.slice(yrIdx, i).reduce((acc, g) => acc * (1 + g), 1)
        : 1;

      revenueByYear[i] += addon.revenue * addOnGrowth + (yearsOwned > 0 ? addon.synergy_revenue : 0);
      ebitdaByYear[i] += addon.revenue * addOnGrowth * addon.ebitda_margin + (yearsOwned > 0 ? addon.synergy_cost : 0);
      if (yearsOwned > 0) {
        synergyRevByYear[i] += addon.synergy_revenue;
        costSynByYear[i] += addon.synergy_cost;
      }
    }
  }

  return {
    revenue_by_year: revenueByYear,
    ebitda_by_year: ebitdaByYear,
    equity_deployed_by_year: equityByYear,
    debt_added_by_year: debtByYear,
    integration_cost_by_year: integrationByYear,
    synergy_revenue_by_year: synergyRevByYear,
    cost_synergies_by_year: costSynByYear,
  };
}

/**
 * Build synthetic debt tranches representing acquisition financing for add-ons.
 * One bullet tranche per year with incremental add-on debt, drawn in that year
 * (via draw_year_index) and carried to exit (captured in exit net debt). Rate
 * fields are copied from a reference senior tranche so the cost of the
 * acquisition financing flows through interest, leverage, coverage and sweep.
 */
export function buildAddOnDebtTranches(state: ModelState, impact: AddOnImpact): DebtTranche[] {
  const hp = state.exit.holding_period;
  const ref = state.debt_tranches.find((t) => t.tranche_type === 'senior') ?? state.debt_tranches[0];
  const out: DebtTranche[] = [];

  for (let yrIdx = 0; yrIdx < impact.debt_added_by_year.length; yrIdx++) {
    const amt = impact.debt_added_by_year[yrIdx];
    if (amt <= 0) continue;
    out.push({
      name: `Add-on Debt (Yr ${yrIdx + 1})`,
      tranche_type: 'senior',
      principal: amt,
      interest_rate: ref ? ref.interest_rate : 0.08,
      rate_type: ref ? ref.rate_type : 'fixed',
      base_rate: ref ? ref.base_rate : 0,
      spread: ref ? ref.spread : 0,
      amortization_type: 'bullet',
      // All-zero schedule: principal carries to exit and is captured in exit net debt
      // (repaid from exit proceeds), matching the bullet-at-exit convention.
      amortization_schedule: Array(hp).fill(0),
      pik_rate: 0,
      cash_interest: true,
      commitment_fee: 0,
      floor: ref ? ref.floor : 0,
      cash_sweep_pct: 0,
      draw_year_index: yrIdx,
      _synthetic_addon: true,
    });
  }
  return out;
}

/**
 * Apply add-on impact to a model state for a single compute pass:
 *   1. strip any pre-existing synthetic tranches (idempotent across recalcs),
 *   2. inject add-on revenue into acquisition_revenue,
 *   3. transiently append synthetic acquisition-debt tranches.
 *
 * Returns the impact plus the original (non-synthetic) tranche list. Callers
 * MUST restore `state.debt_tranches = originalTranches` after computing the
 * schedule so the synthetic tranches never leak into the input panel, scenario
 * clones, or persisted state.
 */
export function injectAddOns(state: ModelState): { impact: AddOnImpact; originalTranches: DebtTranche[] } {
  const baseTranches = state.debt_tranches.filter((t) => !t._synthetic_addon);
  state.debt_tranches = baseTranches;

  const impact = computeAddOnImpact(state);
  state.revenue.acquisition_revenue = impact.revenue_by_year;

  const synthetic = buildAddOnDebtTranches(state, impact);
  if (synthetic.length) {
    state.debt_tranches = [...baseTranches, ...synthetic];
  }
  return { impact, originalTranches: baseTranches };
}

/** Remove any synthetic add-on tranches (defensive idempotency guard). */
export function stripSyntheticAddOnTranches(state: ModelState): void {
  if (state.debt_tranches.some((t) => t._synthetic_addon)) {
    state.debt_tranches = state.debt_tranches.filter((t) => !t._synthetic_addon);
  }
}

File: backend/engine/add_ons.py
"""Add-on acquisition engine (Python mirror of lib/engine/addOns.ts) — D.

Injects bolt-on revenue/EBITDA into the projection (via acquisition_revenue) and
bolt-on acquisition debt into the debt schedule (as synthetic tranches drawn in the
acquisition year). The sponsor-equity portion of equity/mixed-funded bolt-ons is
surfaced via ``equity_deployed_by_year`` and charged in returns (see returns.py).

Previously Python had no add-on engine at all (a documented divergence); this brings
it to parity with the TypeScript source of truth.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from backend.models.state import ModelState
from backend.models.debt import DebtTranche


@dataclass
class AddOnImpact:
    revenue_by_year: list[float] = field(default_factory=list)
    ebitda_by_year: list[float] = field(default_factory=list)
    equity_deployed_by_year: list[float] = field(default_factory=list)
    debt_added_by_year: list[float] = field(default_factory=list)
    integration_cost_by_year: list[float] = field(default_factory=list)
    synergy_revenue_by_year: list[float] = field(default_factory=list)
    cost_synergies_by_year: list[float] = field(default_factory=list)


def compute_add_on_impact(state: ModelState) -> AddOnImpact:
    hp = state.exit.holding_period
    add_ons = state.add_on_acquisitions or []

    revenue_by_year = [0.0] * hp
    ebitda_by_year = [0.0] * hp
    equity_by_year = [0.0] * hp
    debt_by_year = [0.0] * hp
    integration_by_year = [0.0] * hp
    synergy_rev_by_year = [0.0] * hp
    cost_syn_by_year = [0.0] * hp

    growth = state.revenue.growth_rates or []

    for addon in add_ons:
        acq_year = addon.year  # 1-indexed
        if acq_year < 1 or acq_year > hp:
            continue

        purchase_price = addon.revenue * addon.ebitda_margin * addon.purchase_multiple
        if addon.funding == "debt":
            debt_portion = purchase_price
        elif addon.funding == "equity":
            debt_portion = 0.0
        else:  # mixed
            debt_portion = purchase_price * addon.debt_pct
        equity_portion = purchase_price - debt_portion

        yr_idx = acq_year - 1
        equity_by_year[yr_idx] += equity_portion
        debt_by_year[yr_idx] += debt_portion
        integration_by_year[yr_idx] += addon.integration_cost

        for i in range(yr_idx, hp):
            years_owned = i - yr_idx
            # Add-on revenue grows at the base business growth rate.
            add_on_growth = 1.0
            if years_owned > 0 and i < len(growth):
                for g in growth[yr_idx:i]:
                    add_on_growth *= 1 + g

            # Synergy ramp (P4-12): phase in linearly over synergy_ramp_years;
            # absent ⇒ full synergy from the year after acquisition.
            ramp = addon.synergy_ramp_years
            if ramp and ramp > 0:
                synergy_fraction = min(1.0, years_owned / ramp)
            else:
                synergy_fraction = 1.0 if years_owned > 0 else 0.0

            revenue_by_year[i] += addon.revenue * add_on_growth + addon.synergy_revenue * synergy_fraction
            ebitda_by_year[i] += addon.revenue * add_on_growth * addon.ebitda_margin + addon.synergy_cost * synergy_fraction
            if synergy_fraction > 0:
                synergy_rev_by_year[i] += addon.synergy_revenue * synergy_fraction
                cost_syn_by_year[i] += addon.synergy_cost * synergy_fraction

    return AddOnImpact(
        revenue_by_year=revenue_by_year,
        ebitda_by_year=ebitda_by_year,
        equity_deployed_by_year=equity_by_year,
        debt_added_by_year=debt_by_year,
        integration_cost_by_year=integration_by_year,
        synergy_revenue_by_year=synergy_rev_by_year,
        cost_synergies_by_year=cost_syn_by_year,
    )


def build_add_on_debt_tranches(state: ModelState, impact: AddOnImpact) -> list[DebtTranche]:
    """Synthetic bullet tranches for add-on acquisition debt — one per year with
    incremental debt, drawn in that year (draw_year_index) and carried to exit."""
    hp = state.exit.holding_period
    ref = next((t for t in state.debt_tranches if t.tranche_type == "senior"), None)
    if ref is None and state.debt_tranches:
        ref = state.debt_tranches[0]

    out: list[DebtTranche] = []
    for yr_idx, amt in enumerate(impact.debt_added_by_year):
        if amt <= 0:
            continue
        out.append(DebtTranche(
            name=f"Add-on Debt (Yr {yr_idx + 1})",
            tranche_type="senior",
            principal=amt,
            interest_rate=ref.interest_rate if ref else 0.08,
            rate_type=ref.rate_type if ref else "fixed",
            base_rate=ref.base_rate if ref else 0.0,
            spread=ref.spread if ref else 0.0,
            amortization_type="bullet",
            amortization_schedule=[0.0] * hp,
            pik_rate=0.0,
            cash_interest=True,
            commitment_fee=0.0,
            floor=ref.floor if ref else 0.0,
            cash_sweep_pct=0.0,
            draw_year_index=yr_idx,
            synthetic_addon=True,
        ))
    return out


def strip_synthetic_add_on_tranches(state: ModelState) -> None:
    if any(getattr(t, "synthetic_addon", False) for t in state.debt_tranches):
        state.debt_tranches = [t for t in state.debt_tranches if not getattr(t, "synthetic_addon", False)]


def inject_add_ons(state: ModelState) -> tuple[AddOnImpact, list[DebtTranche]]:
    """Strip stale synthetic tranches, inject add-on revenue, and append synthetic
    acquisition-debt tranches. Returns (impact, original non-synthetic tranches);
    callers MUST restore state.debt_tranches = original after computing the schedule."""
    base_tranches = [t for t in state.debt_tranches if not getattr(t, "synthetic_addon", False)]
    state.debt_tranches = base_tranches

    impact = compute_add_on_impact(state)
    state.revenue.acquisition_revenue = impact.revenue_by_year

    synthetic = build_add_on_debt_tranches(state, impact)
    if synthetic:
        state.debt_tranches = base_tranches + synthetic
    return impact, base_tranches
